psi4/driver/p4util/solvers.py

```diff
@@ -304,3 +308,591 @@ def extrapolate(self, out=None):
+
+    vector_dot(X, Y) -> a:
+       X : single `vector`
+       Y : single `vector`
+       a : float, the dot product  (X x Y)
+
+    vector_axpy(a, X, Y) - > Y':
+       a : float
+       X : single `vector`
+       Y : single `vector`
+       Y': single `vector`
+         Y' = Y + a*X (it is assumed that the vector passed as Y is modified. This may not be true)
+
+    vector_scale(a, X) -- > X':
+       a : float
+       X : single `vector`
+       X': single `vector`
+          X' = a * X (it is assumed that the vector passed as X is modified. This may not be true)
+
+    vector_copy(X) -- > X':
+       X : single `vector`
+       X': single `vector`
+         a copy of X
+
+     new_vector() -- > X:
+       X : single `vector`
+         A new object that has the same shape as `vector` like quantities.
+
+
+     ..note:: The `vector` referred to here is intentionally vague, the solver does not care what it is and only
+              holds individual or sets of them. In fact an individual `vector` could be split across two elements in a list,
+              such as for different spin. Whatever data type is used and individual vector should be a single element in a list such that
+              len(list) returns the number of vector-like objects.
+    """
+    nk = nroot
+
+    iter_info = {
+        "count": 0,
+        "res_norm": np.zeros((nk)),
+        "val": np.zeros((nk)),
+        "delta_val": np.zeros((nk)),
+        # conv defaults to true, and will be flipped when a non-conv root is hit
+        "done": True,
+        "nvec": 0,
+        "collapse": False,
+        "product_count": 0,
+    }
+
+    print_name = "DavidsonSolver"
+    title_lines = ["Generalized Davidson Solver", "By Ruhee Dcunha"]
+    max_ss_size = max_vecs_per_root * nk
+
+    _diag_print_heading(title_lines, print_name, max_ss_size, nroot, r_tol, e_tol, maxiter, verbose)
+
+    vecs = guess
+    stats = []
+    best_eigvecs = []
+    best_eigvals = []
+    while iter_info['count'] < maxiter:
+        # increment iteration/ save old vals
+        iter_info['count'] += 1
+        old_vals = iter_info['val'].copy()
+        # reset flags
+        iter_info['collapse'] = False
+        iter_info['done'] = True
+
+        l = len(vecs)
+        iter_info['nvec'] = l
+        if l >= max_ss_size:
+            iter_info['collapse'] = True
+
+        Ax, nprod = engine.compute_products(vecs)
+        iter_info['product_count'] += nprod
+        G = np.zeros((l, l))
+        for i in range(l):
+            for j in range(i):
+                G[i, j] = G[j, i] = engine.vector_dot(vecs[i], Ax[j])
+            G[i, i] = engine.vector_dot(vecs[i], Ax[i])
+
+        _print_array("SS transformed A", G, verbose)
+
+        lam, alpha = np.linalg.eigh(G)
+
+        _print_array("SS eigenvectors", alpha, verbose)
+        _print_array("SS eigenvalues", lam, verbose)
+
+        # remove zeros/negatives
+        alpha = alpha[:, lam > 0.0]
```

Comment:
Maybe build `lam > 0.0` as a temporary mask... do you need to be careful to check for something like `-1e-15`?